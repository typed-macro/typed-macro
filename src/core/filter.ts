import { posix, win32 } from 'path'
import pm from 'picomatch'

export type FilterPattern = RegExp | string | Array<RegExp | string>

export type FilterOptions = {
  /**
   * Regexp or glob patterns to include matches, applied after
   * {@link FilterOptions.exclude}.
   *
   * The default value is `/\.[jt]sx?$/`.
   *
   * Once this option is defined, the default one will be overwritten.
   */
  include?: FilterPattern
  /**
   * Regexp or glob patterns to exclude matches, applied before
   * {@link FilterOptions.include}.
   *
   * The default value is `/node_modules/`.
   *
   * Once this option is defined, the default one will be overwritten.
   */
  exclude?: FilterPattern
}

export function createFilter({
  include = /\.[jt]sx?$/,
  exclude = /node_modules/,
}: FilterOptions) {
  const getMatcher = (pattern: string | RegExp) => {
    if (pattern instanceof RegExp) return pattern
    const fn = pm(pattern, { dot: true })
    return { test: (id: string) => fn(id) }
  }
  const ensureArray = (v: unknown) => (v ? (Array.isArray(v) ? v : [v]) : [])

  const includeMatchers = ensureArray(include).map(getMatcher)
  const hasIncludeRule = includeMatchers.length > 0
  const excludeMatchers = ensureArray(exclude).map(getMatcher)

  return (id: string) => {
    const pathId = id.split(win32.sep).join(posix.sep)
    for (let i = 0; i < excludeMatchers.length; ++i) {
      if (excludeMatchers[i].test(pathId)) return false
    }

    for (let i = 0; i < includeMatchers.length; ++i) {
      if (includeMatchers[i].test(pathId)) return true
    }

    return !hasIncludeRule
  }
}
