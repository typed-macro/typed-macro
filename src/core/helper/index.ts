export { BABEL_TOOLS, Babel } from './babel'
export { ImportOption } from './import'
export { getPathHelper, PathHelper } from './path'
export { getStateHelper, State, StateHelper } from './state'
export { getTransformHelper, TransformHelper } from './transform'
