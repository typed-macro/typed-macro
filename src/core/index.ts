export { Runtime, RuntimeOptions } from './runtime'
export {
  createTransformer,
  Transformer,
  TransformerOptions,
} from './transformer'
export {
  Macro,
  MacroContext,
  MacroHelper,
  MacroHandler,
  Exportable,
  NamespacedExportable,
  NormalizedExports,
  normalizeExports,
} from './types'
