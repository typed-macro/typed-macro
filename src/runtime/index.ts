import {
  NamespacedMacros,
  NamespacedModules,
  NamespacedTypes,
  NormalizedExports,
  renderTypes,
} from '@/runtime/types'
import {
  createTransformer,
  Transformer,
  TransformerOptions,
} from '@/runtime/transformer'
import { findDuplicatedItem } from '@/common'
import { mkdir, writeFile } from 'fs/promises'
import { dirname } from 'path'

export type RuntimeOptions = {
  transformer: TransformerOptions
}

export class Runtime {
  private macros: NamespacedMacros = Object.create(null)
  private modules: NamespacedModules = Object.create(null)
  private types: NamespacedTypes = Object.create(null)

  private macrosNamespaces: string[] = []
  private modulesNamespaces: string[] = []

  private devMode = false

  private readonly transformer: Transformer

  constructor(options: RuntimeOptions) {
    this.transformer = createTransformer(options.transformer)
  }

  setDevMode(dev = true) {
    this.devMode = dev
  }

  register({ macros, modules, types }: NormalizedExports) {
    {
      const duplicated = findDuplicatedItem(
        Object.keys(types),
        Object.keys(this.types)
      )
      if (duplicated) throw new Error(`duplicated namespace '${duplicated}'.`)
    }
    {
      Object.keys(macros).forEach((ns) => {
        const mem = Object.create(null)
        macros[ns].forEach((m) => {
          if (mem[m.name]) {
            throw new Error(
              `a macro with name '${m.name}' in '${ns}' already existed`
            )
          }
          mem[m.name] = 1
        })
      })
    }
    Object.assign(this.macros, macros)
    this.macrosNamespaces = Object.keys(this.macros)
    Object.assign(this.modules, modules)
    this.modulesNamespaces = Object.keys(this.modules)
    Object.assign(this.types, types)
  }

  handleLoad(id: string) {
    if (this.macrosNamespaces.includes(id)) return 'export {}'
    if (this.modulesNamespaces.includes(id)) return this.modules[id]
  }

  handleResolveId(id: string) {
    if (
      this.macrosNamespaces.includes(id) ||
      this.modulesNamespaces.includes(id)
    )
      return id
  }

  handleTransform(code: string, id: string, ssr = false) {
    if (/\.[jt]sx?$/.test(id))
      return this.transformer(
        {
          code,
          id,
          ssr,
          dev: this.devMode,
        },
        this.macros
      )
  }

  get container() {
    return {
      macros: this.macros,
      modules: this.modules,
      types: this.types,
    }
  }

  async generateDts(path: string) {
    await mkdir(dirname(path), { recursive: true })
    await writeFile(path, renderTypes(this.types))
  }
}
