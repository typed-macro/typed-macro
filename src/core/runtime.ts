import {
  createTypeRenderer,
  TypeRenderer,
  TypeRendererOptions,
} from '@/core/typeRenderer'
import {
  createTransformer,
  Transformer,
  TransformerOptions,
} from '@/core/transformer'
import {
  NamespacedMacros,
  NamespacedModules,
  NamespacedTypes,
  NormalizedExports,
  validateMacros,
} from './exports'
import { DeepPartial, findDuplicatedItem } from '@/common'
import { createFilter, FilterOptions } from './filter'

export type RuntimeOptions = {
  /**
   * The options for transformer.
   */
  transformer: TransformerOptions
  /**
   * The options for typeRenderer.
   */
  typeRenderer: Pick<TypeRendererOptions, 'typesPath'>
  /**
   * The options that control the filtering of which files
   * require transformations to be applied.
   */
  filter: FilterOptions
}

export type Attachable = {
  exports: NormalizedExports
  options?: DeepPartial<RuntimeOptions>
}

export class Runtime {
  private macros: NamespacedMacros = Object.create(null)
  private modules: NamespacedModules = Object.create(null)
  private types: NamespacedTypes = Object.create(null)

  private macrosNamespaces: string[] = []
  private modulesNamespaces: string[] = []

  private devMode = false

  constructor(
    private readonly _options: RuntimeOptions,
    defaultExports?: NormalizedExports
  ) {
    if (defaultExports) this.addExports(defaultExports)
  }

  setDevMode(dev = true) {
    this.devMode = dev
  }

  addExports({ macros, modules, types }: NormalizedExports) {
    assertNoDuplicatedNamespace(Object.keys(this.types), Object.keys(types))
    assertNoDuplicatedNamespace(this.macrosNamespaces, Object.keys(macros))
    assertNoDuplicatedNamespace(this.modulesNamespaces, Object.keys(modules))
    validateMacros(macros)
    Object.assign(this.macros, macros)
    this.macrosNamespaces = Object.keys(this.macros)
    Object.assign(this.modules, modules)
    this.modulesNamespaces = Object.keys(this.modules)
    Object.assign(this.types, types)
  }

  handleLoad(id: string) {
    if (this.macrosNamespaces.includes(id)) return 'export {}'
    if (this.modulesNamespaces.includes(id)) return this.modules[id]
  }

  handleResolveId(id: string) {
    if (
      this.macrosNamespaces.includes(id) ||
      this.modulesNamespaces.includes(id)
    )
      return id
  }

  private _transformer?: Transformer
  private _typeRenderer?: TypeRenderer
  private _filter?: (id: string) => boolean

  get attachable() {
    return {
      options: this._options,
      exports: {
        macros: this.macros,
        modules: this.modules,
        types: this.types,
      },
    }
  }

  get filter() {
    return this._filter || (this._filter = createFilter(this._options.filter))
  }

  get transformer() {
    return (
      this._transformer ||
      (this._transformer = createTransformer(this._options.transformer))
    )
  }

  get typeRenderer() {
    return (
      this._typeRenderer ||
      (this._typeRenderer = createTypeRenderer({
        types: this.types,
        typesPath: this._options.typeRenderer.typesPath,
      }))
    )
  }

  handleTransform(code: string, filepath: string, ssr = false) {
    if (this.filter(filepath))
      return this.transformer(
        {
          code,
          filepath,
          ssr,
          dev: this.devMode,
        },
        this.macros
      )
  }

  mergeOptions(options: DeepPartial<RuntimeOptions>) {
    if (options.transformer) {
      // reset transformer
      this._transformer = undefined
      // merge transformer#parserPlugins
      if (options.transformer.parserPlugins?.length) {
        this._options.transformer.parserPlugins = Array.from(
          new Set([
            ...(this._options.transformer.parserPlugins || []),
            ...options.transformer.parserPlugins,
          ])
        )
      }
    }
  }

  attach({ exports, options }: Attachable) {
    // merge exports
    this.addExports(exports)
    // merge options
    options && this.mergeOptions(options)
  }
}

export function assertNoDuplicatedNamespace(ns1: string[], ns2: string[]) {
  const duplicated = findDuplicatedItem(ns1, ns2)
  if (duplicated) throw new Error(`duplicated namespace '${duplicated}'.`)
}
