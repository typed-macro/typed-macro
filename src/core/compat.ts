/*
 * This is a compatibility control mechanism based on version number,
 * to ensure the compatibility between macro and runtime.
 */

import { Macro } from '@/core/macro'
import { WithVersion } from '@/common'

/*
 * The current Macro version.
 *
 * It represents whether the call-context passed by the Runtime to the Macro is compatible.
 * This involves not only the structure of the call-context itself,
 * but also the structure of the AST between different versions of Babel,
 * for AST is generated in Runtime but transformed in Macros.
 *
 * @deps {@link RawMacroCall}, {@link Babel}
 */
export const MACRO_VERSION = 0

export function versionedMacro(v: Macro): Macro {
  ;(v as WithVersion<Macro>).$__version = MACRO_VERSION
  return v
}

export function isMacroCompatible(v: Macro) {
  return (v as WithVersion<Macro>).$__version === MACRO_VERSION
}
