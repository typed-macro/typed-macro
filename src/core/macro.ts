import { CallExpression } from '@babel/types'
import { NodePath } from '@babel/traverse'
import { Babel, PathHelper, StateHelper, TransformHelper } from '@/core/helper'

export type MacroContext = {
  /**
   * The file path of the module currently being handled.
   */
  filepath: string
  /**
   * The original source code of the module currently being handled.
   */
  code: string
  /**
   * The arguments nodes of the call-macro expression currently being handled.
   */
  args: CallExpression['arguments']
  /*
   * The NodePath of the call-macro expression currently being handled.
   */
  path: NodePath
  /*
   * Whether in ssr mode.
   */
  ssr: boolean
}

export type MacroHelper = TransformHelper & PathHelper & StateHelper

export type MacroHandler = (
  /**
   * Context about the macro caller.
   */
  ctx: MacroContext,
  /**
   * Babel tools to help operate node paths and nodes.
   */
  babel: Readonly<Babel>,
  /**
   * Wrappers upon babel tools to simplify operations.
   */
  helper: Readonly<MacroHelper>
) => void

export type Macro = {
  name: string
  apply: MacroHandler
}

export type MacroMeta = {
  signatures: {
    comment?: string
    signature: string
  }[]
  types: string[]
}

type InternalMacro = Macro & {
  __types: string
}

export function macro(
  name: string,
  meta: MacroMeta,
  handler: MacroHandler
): Macro {
  return {
    name,
    apply: handler,
    __types: renderMetaType(name, meta),
  } as InternalMacro
}

export function isMacro(o: unknown): o is InternalMacro {
  return !!(o as InternalMacro).__types
}

export function renderMetaType(name: string, meta: MacroMeta) {
  return [
    meta.types.join('\n'),
    meta.signatures
      .map((s) =>
        s.comment
          ? `  /** ${s.comment} **/
  export function ${name}${s.signature}`
          : `  export function ${name}${s.signature}`
      )
      .join('\n'),
  ]
    .filter((t) => !!t)
    .join('\n')
}
