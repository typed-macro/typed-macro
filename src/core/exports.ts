import { isMacro, Macro } from '@/core/macro'
import { isMacroCompatible } from '@/core/compat'

export type NamespacedMacros = { [namespace: string]: Macro[] }

export type NamespacedModules = { [namespace: string]: string }

export type NamespacedTypes = {
  [namespace: string]: {
    moduleScope: string[]
    macroScope: string[]
  }
}

export type Exportable = (
  | {
      /**
       * as macros
       */
      macros: Macro[]
    }
  | {
      /**
       * as module, code will be used as source code of the module
       */
      code: string
    }
) & {
  /**
   * Type definitions, will be written to d.ts.
   *
   * e.g.
   * ```typescript
   * { exports: { '@macros': { customTypes: `type A = string` } } }
   * ```
   * will generate
   * ```typescript
   * declare module '@macros' {
   *   type A = string
   * }
   * ```
   */
  customTypes?: string
}

export type NamespacedExportable = { [namespace: string]: Exportable }

export type NormalizedExports = {
  macros: NamespacedMacros
  modules: NamespacedModules
  types: NamespacedTypes
}

export function normalizeExports(
  exports: NamespacedExportable
): NormalizedExports {
  const macros: NamespacedMacros = Object.create(null)
  const modules: NamespacedModules = Object.create(null)
  const types: NamespacedTypes = Object.create(null)
  Object.keys(exports).forEach((ns) => {
    const item = exports[ns]
    types[ns] = {
      moduleScope: item.customTypes ? [item.customTypes] : [],
      macroScope: [],
    }
    if ('code' in item) {
      modules[ns] = item.code
    } else {
      macros[ns] = item.macros
      types[ns].macroScope = item.macros.map((m) => {
        if (!isMacro(m)) throw new Error(`'${m.name}' is not a macro`)
        return m.__types
      })
    }
  })

  validateMacros(macros)

  return {
    macros,
    modules,
    types,
  }
}

// ensure compatible and no name conflict
export function validateMacros(macros: NamespacedMacros) {
  Object.keys(macros).forEach((ns) => {
    const mem = Object.create(null)
    macros[ns].forEach((m) => {
      if (!isMacroCompatible(m))
        throw new Error(`macro '${m.name}' in '${ns}' is incompatible`)

      if (mem[m.name])
        throw new Error(`macro '${m.name}' in '${ns}' is duplicated`)

      mem[m.name] = 1
    })
  })
}
