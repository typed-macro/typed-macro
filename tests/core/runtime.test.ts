import { assertNoDuplicatedNamespace, Runtime } from '@/core/runtime'
import { macroSerializer, mockMacro, mockRuntime } from '../testutils'

expect.addSnapshotSerializer(macroSerializer)

describe('assertNoDuplicatedNamespace()', () => {
  it('should work', () => {
    expect(() => assertNoDuplicatedNamespace(['a'], ['b'])).not.toThrow()
    expect(() => assertNoDuplicatedNamespace(['a', 'b'], ['c', 'b'])).toThrow()
  })
})

describe('Runtime', () => {
  let runtime: Runtime
  beforeEach(() => {
    runtime = mockRuntime()
  })

  it('should work with setDevMode()', () => {
    runtime.setDevMode()
    expect((runtime as any).devMode).toBe(true)
  })

  it('should work with addExports() and .exports', () => {
    runtime.addExports({
      macros: { '@m1': [mockMacro('m')] },
      modules: { '@u1': 'export const a = 1' },
      types: {},
    })
    expect(runtime.exports).toMatchSnapshot()
    runtime.addExports({
      macros: { '@m2': [mockMacro('m')] },
      modules: { '@u2': 'export const a = 1' },
      types: {},
    })
    expect(runtime.exports).toMatchSnapshot()
    expect(() => {
      runtime.addExports({
        macros: { '@m2': [] },
        modules: {},
        types: {},
      })
    }).toThrow()
    expect(() => {
      runtime.addExports({
        macros: {
          '@m3': [mockMacro('m'), mockMacro('m')],
        },
        modules: {},
        types: {},
      })
    }).toThrow()
  })

  it('should work with mergeOptions() and .options', () => {
    expect(runtime.options.transformer).toEqual({})
    runtime.mergeOptions({})
    expect(runtime.options.transformer).toEqual({})
    runtime.mergeOptions({ transformer: { parserPlugins: ['topLevelAwait'] } })
    expect(runtime.options.transformer.parserPlugins).toEqual(['topLevelAwait'])
    runtime.mergeOptions({
      transformer: { parserPlugins: ['topLevelAwait', 'decorators'] },
    })
    expect(runtime.options.transformer.parserPlugins).toEqual([
      'topLevelAwait',
      'decorators',
    ])
    runtime.mergeOptions({
      transformer: { maxRecursions: 5 },
    })
    expect(runtime.options.transformer.maxRecursions).toBeUndefined()
  })

  it('should work with handleLoad/handleResolveId()', () => {
    runtime.addExports({
      macros: { '@m': [mockMacro('m')] },
      modules: { '@u': 'export const a = 1' },
      types: {},
    })
    expect(runtime.handleLoad('@m')).toMatchSnapshot()
    expect(runtime.handleLoad('@u')).toMatchSnapshot()
    expect(runtime.handleLoad('@n')).toBeUndefined()

    expect(runtime.handleResolveId('@m')).toMatchSnapshot()
    expect(runtime.handleResolveId('@u')).toMatchSnapshot()
    expect(runtime.handleResolveId('@n')).toBeUndefined()
  })

  it('should work with handleTransform()', () => {
    runtime.addExports({
      macros: {
        '@m': [
          mockMacro('m', ({ path }) => {
            path.remove()
          }),
        ],
      },
      modules: {},
      types: {},
    })
    // no macro
    expect(
      runtime.handleTransform(`console.log('hello')`, 'test.js')
    ).toBeUndefined()
    // not valid filename
    expect(
      runtime.handleTransform(`import {m} from '@m'; m()`, 'a')
    ).toBeUndefined()

    expect(
      runtime.handleTransform(
        `import {m} from '@m'; m(); console.log('helloworld')`,
        'test.js'
      )
    ).toMatchSnapshot()
  })

  it('should work with typeRenderer()', () => {
    runtime.addExports({
      macros: {},
      modules: {},
      types: {
        '@t': {
          moduleScope: ['type A = string'],
          macroScope: [],
        },
      },
    })

    expect(runtime.typeRenderer.render()).toMatchSnapshot()

    runtime.addExports({
      macros: {},
      modules: {},
      types: {
        '@n': {
          moduleScope: ['type B = string'],
          macroScope: [`export function a(): void`],
        },
      },
    })

    // update after register new
    expect(runtime.typeRenderer.render()).toMatchSnapshot()

    const update = runtime.typeRenderer.append('@l', 'type C = string')
    expect(runtime.typeRenderer.render()).toMatchSnapshot()
    update('type G = number')
    expect(runtime.typeRenderer.render()).toMatchSnapshot()
  })
})
