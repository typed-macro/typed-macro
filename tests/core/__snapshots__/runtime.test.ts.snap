// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Runtime should work with addExports() and .exports 1`] = `
Object {
  "macros": Object {
    "@m1": Array [
      Macro { "name": "m", "__types": "" },
    ],
  },
  "modules": Object {
    "@u1": "export const a = 1",
  },
  "types": Object {},
}
`;

exports[`Runtime should work with addExports() and .exports 2`] = `
Object {
  "macros": Object {
    "@m1": Array [
      Macro { "name": "m", "__types": "" },
    ],
    "@m2": Array [
      Macro { "name": "m", "__types": "" },
    ],
  },
  "modules": Object {
    "@u1": "export const a = 1",
    "@u2": "export const a = 1",
  },
  "types": Object {},
}
`;

exports[`Runtime should work with handleLoad/handleResolveId() 1`] = `"export {}"`;

exports[`Runtime should work with handleLoad/handleResolveId() 2`] = `"export const a = 1"`;

exports[`Runtime should work with handleLoad/handleResolveId() 3`] = `"@m"`;

exports[`Runtime should work with handleLoad/handleResolveId() 4`] = `"@u"`;

exports[`Runtime should work with handleTransform() 1`] = `"console.log('helloworld');"`;

exports[`Runtime should work with typeRenderer() 1`] = `
"declare module '@t' {
type A = string
}"
`;

exports[`Runtime should work with typeRenderer() 2`] = `
"declare module '@t' {
type A = string
}
declare module '@n' {
type B = string
export function a(): void
}"
`;

exports[`Runtime should work with typeRenderer() 3`] = `
"declare module '@t' {
type A = string
}
declare module '@n' {
type B = string
export function a(): void
}
declare module '@l' {
type C = string
}"
`;

exports[`Runtime should work with typeRenderer() 4`] = `
"declare module '@t' {
type A = string
}
declare module '@n' {
type B = string
export function a(): void
}
declare module '@l' {
type G = number
}"
`;
